src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/H2Engine.java

```diff
@@ -533,13 +533,17 @@ protected PreparedStatement getPreparedStatementForPersist(final boolean useAuto
     protected synchronized long doPersist(final PreparedStatement ps,
                                           final MappedEntity me,
                                           final boolean useAutoInc,
-                                          int lastBindPosition) throws Exception {
+                                          final int lastBindPosition) throws Exception {
         ps.execute();
 
         if (useAutoInc) {
-            try (final ResultSet generatedKeys = ps.getGeneratedKeys()) {
-                if (generatedKeys.next()) {
-                    return generatedKeys.getLong(1);
+            for (final DbColumn column : me.getEntity().getColumns()) {
+                if ((column.isAutoInc())) {
+                    try (final ResultSet generatedKeys = ps.getGeneratedKeys()) {
+                        if (generatedKeys.next()) {
+                            return generatedKeys.getLong(column.getName());
+                        }
+                    }
```

Comment:
I don't know the performance of a cycle for versus something like `final DbColumn column = me.getEntity().getColumns().stream().filter(c -> c.isAutoInc()).findFirst().get();`.
This is an important method that we need to try not to introduce cyclomatic complexity.

Reply:
Actually, MappedEntity has `.getAutoIncColumn()`.
You just need to change the `return generatedKeys.getLong(1);` to `return generatedKeys.getLong(me.getAutoIncColumn());`

Reply:
For H2 `getAutoIncColumn()` is not being set :disappointed: 
We can either find a wait to set it on the binding of the statement on `entityToPreparedStatement()` otherwise there's always the need to go through the columns.

I agree in the stream by the way :smile:

Reply:
Probably we need to set the column as auto increment when creating it... https://www.baeldung.com/jdbc-returning-generated-keys

Reply:
This new fix already creates a new `MappedEntity` with the last column found that contains `autoinc` values set as the `IncColumn` in the `MappedEntity`.